### ISSUE TITLE
C# Exported Array<Node> is null

### ISSUE THREAD

**Issue by LasseSchnepel** [godotengine/godot#97190]
### Tested versions

4.3 stable

### System information

Godot v4.3.stable.mono - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1070 Ti (NVIDIA; 31.0.15.3623) - AMD Ryzen 7 2700X Eight-Core Processor (16 Threads)

### Issue description

I have a larger project, that uses C# for all the scripts. Everything works fine when I start the game from the editor. When I export it, I get several errors, that I was able to trace back to that an exported variable is null, instead of being set to the value that is stored in the scene.
This applies to exported variables, that are made of a Godot.Collections.Array<Node>, where <Node> can be Node or any other Script that inherits from Node.
If I replace the Array<Node> with Node[], the export is working fine.

### Steps to reproduce

Unfortunately, I was not able to create a MRP, as the issue did not persist in a minimal project. So unfortunately, I do not know how to reproduce, but I can share the code with one or two people who want to investigate the issue.

### Minimal reproduction project (MRP)

I was not able to create a MRP and don't want to publically share my project, but I can share the code with one or two persons who want to investigate the issue.

**Comment by AThousandShips** [comment]
> but I can share the code with one or two persons who want to investigate the issue.

This will not work, if you are unable to reproduce this in a new project an MRP is going to be necessary, so please try to create one, or this won't be possible to test

**Comment by LasseSchnepel** [comment]
Hi @AThousandShips,
Yes, I understand that. While I was not able to reproduce a MRP from scratch, I now tried to strip everything from my project and I think I have a quiet minimal example now.

While stripping it down and, I realized that there are multiple scenarios, which all behave different:

- If the problematic scene (Level.tscn) is the starting scene:
the editor removes the value of the problematic property from the tscn (checked via texteditor/git). I think this is done because the C# project is not build at that time. I don't think it's good that this is done. For this case, I would even understand that the problematic property is then zero.
- If the problematic scene is not the starting scene:
the problematic scene file (Level.tscn) is unchanged, but the export shows the same problem. If I (after export) open the Level.tscn, the value of the property is removed
- If the problematic scene is not the starting scene:
And the first thing I do in the editor is to open the Level.tscn, it removes the value of my property.
- If the problematic scnee is not the starting scene:
And the first thing I do in the editor is to build the project (not start), then I can open the Level.tscn without losing the value of the property, also an export is then working fine. In this case, the export is also fine if I do it directly after building the project (without opening Level.tscn)

**So, I would sum it up as:**
C# Properties (maybe only Array<Node>) are being removed from the .tscn file if the scene is opened before the project is build for the first time.

I hope this helps, please find the MRP attached:
[mrp.zip](https://github.com/user-attachments/files/17065491/mrp.zip)

**Comment by AThousandShips** [comment]
That should hopefully help! Thank you! Can't test C# projects myself at the moment but I'm sure someone will take a look

**Comment by rsubtil** [comment]
cc @AThousandShips 

I've replicated this bug the following way:
- Set some C# exported array variable (from the MRP, assign something to the Loading Zones)
- Clean the C# build, and reload the project
- Open the previous exported variable, rebuild the project; data will now be missing.

https://github.com/user-attachments/assets/7582d8b7-c2d5-40d0-ab71-ed9aedd249c0

I've compiled Godot with the linked PR fix (https://github.com/godotengine/godot/pull/97864), and the issue seems to be gone; the array contents are preserved after a clean rebuild:

https://github.com/user-attachments/assets/2b1bbabf-88ce-48d9-ba44-a845d83a930c

**Comment by SkaveRat** [comment]
Wow, this took several days of debugging to find.

Initially, the issue was hidden, because local development and exporting worked fine.

But trying to build the project in a (cold cache) CI environment would break, due to Array and Dictionary exports being nulled/emptied.

> C# Properties (maybe only Array) are being removed from the .tscn file if the scene is opened before the project is build for the first time.

This is indeed the culprit.

I can reliably build the project now by first loading the assets and warming up the cache, and then building the solution:

`godot --headless --verbose --editor --quit`

`godot --headless --verbose --editor --quit --build-solutions`

tested with 4.6.3 and 4.7

**Comment by rsubtil** [comment]
@SkaveRat are you able to confirm if https://github.com/godotengine/godot/pull/97864 fixes the issue on your end as well?

**Comment by SkaveRat** [comment]
Looks like the PR has merge conflicts?

Haven't built godot from source yet, but will see if I can get it running later today

**Comment by SkaveRat** [comment]
managed to successfully compile godot

But my c++ skills aren't good enough to reliably fix the merge conflict in the PR

**Comment by rsubtil** [comment]
> But my c++ skills aren't good enough to reliably fix the merge conflict in the PR

No worries, you can try building on the MR branch as is, without fixing the merge conflict. It's version might be an issue though, as it was done a couple months ago, and is on the v4.4 version. Regardless I'll bump the PR to notify of the conflict and hopefully they can fix it.

**Comment by SkaveRat** [comment]
yup, already tried the branch as is. sadly the version is too old for my project

**Comment by kevinlam508** [comment]
@SkaveRat I've made an alternate PR (#120679) if you want to see if that resolves your issue. It's in 4.8, so you should be able to use it with your project.

**Comment by SkaveRat** [comment]
yup, that fixed the issue!

**Comment by BulsonJ** [comment]
I am having this issue in a CI environment too. Exporting locally works fine, in CI I get the same error. Which makes CI essentially useless to me right now.
Did you manage to fix it @SkaveRat ?

**Comment by SkaveRat** [comment]
Doing this in the CI environment works for me: 

https://github.com/godotengine/godot/issues/97190#issuecomment-4798201630

**Comment by BulsonJ** [comment]
> Doing this in the CI environment works for me:
> 
> [#97190 (comment)](https://github.com/godotengine/godot/issues/97190#issuecomment-4798201630)

It worked!!! Thanks. I did try it before but it seems I may have entered it wrong...
